File: python/exercises/practice/unit_test_a49e025e_fdd7_41cd_b02c_4f0ac37aac64/transaction_manager/transaction_manager_test.py
import unittest
import threading
from transaction_manager import TransactionManager

class TestTransactionManager(unittest.TestCase):
    def setUp(self):
        self.tm = TransactionManager()

    def test_begin_transaction(self):
        self.assertTrue(self.tm.begin_transaction(1))
        self.assertFalse(self.tm.begin_transaction(1))  # Duplicate
        self.assertTrue(self.tm.begin_transaction(2))

    def test_register_participant(self):
        self.tm.begin_transaction(1)
        self.assertTrue(self.tm.register_participant(1, 101))
        self.assertFalse(self.tm.register_participant(1, 101))  # Duplicate
        self.assertFalse(self.tm.register_participant(2, 101))  # Non-existent tx

    def test_service_commit(self):
        self.tm.begin_transaction(1)
        self.tm.register_participant(1, 101)
        self.tm.register_participant(1, 102)
        
        self.assertTrue(self.tm.service_commit(1, 101))
        self.assertEqual(self.tm.get_transaction_status(1), "PENDING")
        
        self.assertTrue(self.tm.service_commit(1, 102))
        self.assertEqual(self.tm.get_transaction_status(1), "COMMITTED")

    def test_service_rollback(self):
        self.tm.begin_transaction(1)
        self.tm.register_participant(1, 101)
        self.tm.register_participant(1, 102)
        
        self.assertTrue(self.tm.service_rollback(1, 101))
        self.assertEqual(self.tm.get_transaction_status(1), "ROLLED_BACK")
        self.assertFalse(self.tm.service_commit(1, 102))  # Already rolled back

    def test_concurrent_operations(self):
        results = []
        
        def worker(tx_id, service_id, action):
            if action == 'begin':
                results.append(self.tm.begin_transaction(tx_id))
            elif action == 'register':
                results.append(self.tm.register_participant(tx_id, service_id))
            elif action == 'commit':
                results.append(self.tm.service_commit(tx_id, service_id))
            elif action == 'rollback':
                results.append(self.tm.service_rollback(tx_id, service_id))
        
        threads = []
        threads.append(threading.Thread(target=worker, args=(1, 101, 'begin')))
        threads.append(threading.Thread(target=worker, args=(1, 101, 'register')))
        threads.append(threading.Thread(target=worker, args=(1, 101, 'commit')))
        
        for t in threads:
            t.start()
        for t in threads:
            t.join()
            
        self.assertEqual(results, [True, True, True])
        self.assertEqual(self.tm.get_transaction_status(1), "COMMITTED")

    def test_edge_cases(self):
        # Non-existent transaction
        self.assertEqual(self.tm.get_transaction_status(999), "NON_EXISTENT")
        self.assertFalse(self.tm.service_commit(999, 101))
        self.assertFalse(self.tm.service_rollback(999, 101))
        
        # Commit before registration
        self.tm.begin_transaction(1)
        self.assertFalse(self.tm.service_commit(1, 101))
        
        # Register after commit
        self.tm.begin_transaction(2)
        self.tm.register_participant(2, 201)
        self.assertTrue(self.tm.service_commit(2, 201))
        self.assertFalse(self.tm.register_participant(2, 202))

    def test_multiple_participants(self):
        self.tm.begin_transaction(1)
        for i in range(100, 110):
            self.tm.register_participant(1, i)
        
        # Some commit, some rollback
        for i in range(100, 105):
            self.assertTrue(self.tm.service_commit(1, i))
        self.assertTrue(self.tm.service_rollback(1, 105))
        
        self.assertEqual(self.tm.get_transaction_status(1), "ROLLED_BACK")
        
        # Verify no further operations allowed
        self.assertFalse(self.tm.service_commit(1, 106))
        self.assertFalse(self.tm.register_participant(1, 111))

if __name__ == '__main__':
    unittest.main()

File: java/exercises/practice/unit_test_741d70e7_e907_4f72_832b_33a093ed1226/distributed_tx/src/test/java/TransactionManagerTest.java
import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class TransactionManagerTest {

    private TransactionManager transactionManager;
    private ParticipantService userService;
    private ParticipantService profileService;
    private ParticipantService notificationService;

    @BeforeEach
    public void setup() {
        transactionManager = new TransactionManager();
        userService = new MockParticipantService("UserService");
        profileService = new MockParticipantService("ProfileService");
        notificationService = new MockParticipantService("NotificationService");
    }

    @Test
    public void testSuccessfulTransaction() {
        // Setup a transaction
        String transactionId = transactionManager.beginTransaction();
        transactionManager.registerParticipant(transactionId, userService);
        transactionManager.registerParticipant(transactionId, profileService);
        transactionManager.registerParticipant(transactionId, notificationService);

        // Test transaction commit
        assertTrue(transactionManager.commitTransaction(transactionId), 
                "Transaction should commit successfully when all participants are prepared");

        // Verify state
        assertEquals(TransactionStatus.COMMITTED, transactionManager.getTransactionStatus(transactionId),
                "Transaction status should be COMMITTED after successful commit");
    }

    @Test
    public void testFailedPreparation() {
        // Setup a transaction
        String transactionId = transactionManager.beginTransaction();
        transactionManager.registerParticipant(transactionId, userService);
        transactionManager.registerParticipant(transactionId, new FailingParticipantService("FailingService"));
        transactionManager.registerParticipant(transactionId, notificationService);

        // Test transaction commit (should fail during preparation)
        assertFalse(transactionManager.commitTransaction(transactionId),
                "Transaction should fail when any participant fails to prepare");

        // Verify state
        assertEquals(TransactionStatus.ROLLED_BACK, transactionManager.getTransactionStatus(transactionId),
                "Transaction status should be ROLLED_BACK after failed preparation");
    }

    @Test
    public void testExplicitRollback() {
        // Setup a transaction
        String transactionId = transactionManager.beginTransaction();
        transactionManager.registerParticipant(transactionId, userService);
        transactionManager.registerParticipant(transactionId, profileService);

        // Test transaction rollback
        transactionManager.rollbackTransaction(transactionId);

        // Verify state
        assertEquals(TransactionStatus.ROLLED_BACK, transactionManager.getTransactionStatus(transactionId),
                "Transaction status should be ROLLED_BACK after explicit rollback");
    }

    @Test
    public void testNonExistentTransaction() {
        // Test non-existent transaction
        String invalidTransactionId = "non-existent-transaction";
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            transactionManager.commitTransaction(invalidTransactionId);
        });

        assertTrue(exception.getMessage().contains("Transaction not found"),
                "Exception should indicate that transaction was not found");
    }

    @Test
    public void testCommitWithNoParticipants() {
        // Setup a transaction without participants
        String transactionId = transactionManager.beginTransaction();

        // Test transaction commit
        assertTrue(transactionManager.commitTransaction(transactionId),
                "Transaction with no participants should commit successfully");

        // Verify state
        assertEquals(TransactionStatus.COMMITTED, transactionManager.getTransactionStatus(transactionId),
                "Transaction status should be COMMITTED when there are no participants");
    }

    @Test
    public void testConcurrentTransactions() throws InterruptedException {
        // Create a latch to synchronize thread completion
        final int COUNT = 10;
        CountDownLatch latch = new CountDownLatch(COUNT);
        
        // Start multiple concurrent transactions
        for (int i = 0; i < COUNT; i++) {
            final int index = i;
            new Thread(() -> {
                try {
                    String txId = transactionManager.beginTransaction();
                    ParticipantService service1 = new MockParticipantService("Service1-" + index);
                    ParticipantService service2 = new MockParticipantService("Service2-" + index);
                    
                    transactionManager.registerParticipant(txId, service1);
                    transactionManager.registerParticipant(txId, service2);
                    
                    if (index % 5 == 0) {
                        // Occasionally fail a transaction
                        transactionManager.rollbackTransaction(txId);
                        assertEquals(TransactionStatus.ROLLED_BACK, 
                                transactionManager.getTransactionStatus(txId),
                                "Transaction should be rolled back");
                    } else {
                        boolean result = transactionManager.commitTransaction(txId);
                        assertTrue(result, "Transaction should commit successfully");
                        assertEquals(TransactionStatus.COMMITTED, 
                                transactionManager.getTransactionStatus(txId),
                                "Transaction should be committed");
                    }
                } finally {
                    latch.countDown();
                }
            }).start();
        }
        
        // Wait for all threads to complete
        assertTrue(latch.await(5, TimeUnit.SECONDS), "All transactions should complete within timeout");
    }

    @Test
    public void testTransactionTimeout() {
        // Configure transaction manager with a very short timeout
        TransactionManager timeoutManager = new TransactionManager(100); // 100ms timeout
        
        // Setup a transaction
        String transactionId = timeoutManager.beginTransaction();
        timeoutManager.registerParticipant(transactionId, new SlowParticipantService("SlowService", 500));
        
        // Test transaction commit (should timeout)
        assertFalse(timeoutManager.commitTransaction(transactionId),
                "Transaction should fail when a participant times out");
        
        // Verify state
        assertEquals(TransactionStatus.ROLLED_BACK, timeoutManager.getTransactionStatus(transactionId),
                "Transaction status should be ROLLED_BACK after timeout");
    }

    // Mock implementations for testing
    
    private static class MockParticipantService implements ParticipantService {
        private final String name;
        private final Map<String, String> data = new HashMap<>();
        private final Set<String> preparedTxs = new HashSet<>();
        private final Set<String> committedTxs = new HashSet<>();
        
        public MockParticipantService(String name) {
            this.name = name;
        }
        
        @Override
        public ParticipantStatus prepare(String transactionId) {
            preparedTxs.add(transactionId);
            return ParticipantStatus.PREPARED;
        }
        
        @Override
        public void commit(String transactionId) {
            if (!preparedTxs.contains(transactionId)) {
                throw new IllegalStateException("Cannot commit a transaction that wasn't prepared");
            }
            committedTxs.add(transactionId);
        }
        
        @Override
        public void rollback(String transactionId) {
            preparedTxs.remove(transactionId);
        }
        
        @Override
        public String getName() {
            return name;
        }
    }
    
    private static class FailingParticipantService implements ParticipantService {
        private final String name;
        
        public FailingParticipantService(String name) {
            this.name = name;
        }
        
        @Override
        public ParticipantStatus prepare(String transactionId) {
            return ParticipantStatus.ABORT;
        }
        
        @Override
        public void commit(String transactionId) {
            throw new IllegalStateException("Should never commit a failing service");
        }
        
        @Override
        public void rollback(String transactionId) {
            // No-op for failing service
        }
        
        @Override
        public String getName() {
            return name;
        }
    }
    
    private static class SlowParticipantService implements ParticipantService {
        private final String name;
        private final long delayMs;
        
        public SlowParticipantService(String name, long delayMs) {
            this.name = name;
            this.delayMs = delayMs;
        }
        
        @Override
        public ParticipantStatus prepare(String transactionId) {
            try {
                Thread.sleep(delayMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return ParticipantStatus.PREPARED;
        }
        
        @Override
        public void commit(String transactionId) {
            // No-op for slow service
        }
        
        @Override
        public void rollback(String transactionId) {
            // No-op for slow service
        }
        
        @Override
        public String getName() {
            return name;
        }
    }
}